src/api/hooks.ts

```diff
@@ -0,0 +1,159 @@
+import axios from "axios";
+import {
+  APIError,
+  APIGetHook,
+  APIGetHookPaginated,
+  APIPostHook,
+  RequestState,
+  Issue,
+  User,
+  UserAPIError,
+  Article,
+  Pagination,
+  APIResponse,
+  ArticleContent,
+} from "../shared/types";
+import { useState, useEffect } from "react";
+import { useAuth } from "../auth/AuthProvider";
+import { getApiUrl } from "./api";
+
+const useApiGet = <T, U extends APIError = APIError>(
+  url: string
+): APIGetHook<T, U> => {
+  const [response, setResponse] = useState<T | undefined>(undefined);
+  const [error, setError] = useState<U | undefined>(undefined);
+
+  useEffect(() => {
+    axios.get(url).then((axiosResp) => {
+      if (axiosResp.data.success) setResponse(axiosResp.data.data);
+      else setError(axiosResp.data.error);
+    });
+  }, [url]);
+  return [response, error];
+};
+
+const useApiGetPaginated = <T, U extends APIError = APIError>(
+  url: string
+): APIGetHookPaginated<T, U> => {
+  const [currentUrl, setCurrentUrl] = useState<string>(url);
+  const [resp, error] = useApiGet<Pagination<T>, U>(currentUrl);
+
+  const next = () => {
+    if (resp?.next) {
+      setCurrentUrl(resp.next);
+    }
+  };
+
+  const previous = () => {
+    if (resp?.previous) {
+      setCurrentUrl(resp.previous);
+    }
+  };
+
+  return [
+    {
+      next: resp?.next ? next : null,
+      previous: resp?.previous ? previous : null,
+      page: resp ?? null,
+    },
+    error,
+  ];
+};
+
+const useApiPost = <S, T, U extends APIError = APIError>(
+  url: string
+): APIPostHook<S, T, U> => {
+  const auth = useAuth();
+
+  const [state, setState] = useState<RequestState>(RequestState.NotStarted);
+
+  let headers: { [header: string]: string } = {};
```

Comment:
Can we make this a general type? I see this being a pattern we may encounter a lot in the future.

Reply:
So, I put some more thought into this, and since there are really only four-ish HTTP methods we care about, and and only two of them are the same really, I don't think we should bother trying to abstract this.